### DISCUSSION TITLE
VSCode: Unused imports not highlighted

### DISCUSSION THREAD

**Discussion by dekiesel** [scalameta/metals#4452] [Q&A]
Hi,

metals doesn't highlight unused imports in vscode (probably a config error on my side).

Let's say I have two imports

```
import org.apache.spark.sql.Column
import org.apache.spark.sql.DataFrame

```
`DataFrame` is used, `Column` isn't. Metals doesn't highlight that unused column. 

The "bloop sbt already installed"-checkbox is unticked:

![image](https://user-images.githubusercontent.com/17333827/192979912-609e6b32-e2be-4053-8a33-41306f62eae3.png)

I read somewhere that I should create` projects/build.sbt`, but that seems sbt specific and I assume metals would create that file if needed.

`Organize Imports` does organize the imports but doesn't remove the unused import.

Could somebody tell me how I can get highlighting working?

Thanks!

**Comment by tgodzik** [comment]
Metals uses the compiler option for that, so you would need to add it to scalac options if you want to see unused imports.

The options are -Ywarn-unused (or -Wunused in 2.13).

**Reply by dekiesel** [reply]
Thank you for your help.

I wasn't able to find how to apply this setting globally in the [documentation for vscode](https://scalameta.org/metals/docs/editors/vscode/).

Is it possible have it as the default with metals?

**Reply by tgodzik** [reply]
This needs to be specified within the build tool that you use. It can be later used in scalafix to verify it on CI. For example we do it metals here https://github.com/scalameta/metals/blob/main/build.sbt#L162 together with some other linting options.

Once it's enabled and the build reimported, organize imports will remove the unsued imports automatically.

**Reply by dekiesel** [reply]
So that means I have to define a `build.sbt` file for every one of my projects if I want the highlighting to work?

**Reply by tgodzik** [reply]
You don't have to define a new `build.sbt`, you just need to modify the current build tool you are using. What are you using in your projects?

**Reply by dekiesel** [reply]
Sorry for the stupid questions and thanks for your patience, I am new to scala (as you probably have figured out).

Usually I clone or create a repo (data engineering), make my changes, create and run unit tests (in vscode) and then commit my code. Then a pipeline picks it up and sends it on it's way.

All I did was install the metals extension for vscode (and a local spark "cluster" and everything has been running smoothly. Honestly, I assumed metals would build my code, but I can see a .bloop folder so I guess the answer is "bloop"?

**Reply by tgodzik** [reply]
It's Bloop building the project yeah, but it can also be a number of other build servers. You usually have to import the build from an existing build tool such as Gradle, Mill, Maven or sbt. Some can also be used as the server themselves. By default we will suggest to import your project into Bloop and to do that we run a plugin for the specific build tool. What does exist in your workspace? `build.gradle`, `build.sbt`, `build.sc` or `pom.xml` ? Each of those have a bit different methods of adding scalac options, it might be useful for use to set up a guide for that actually, but it should be simple enough to do it.

 We decided against enabling unused imports by default for everyone and are more "build tool centric" so that whatever you set up is set up for all the users and can be verified in CI.

**Reply by dekiesel** [reply]
We are using maven (and all the projects have a `pom.xml`). I see it is possible to add the option there, but I'd need to do that for every repo and I am the only one in my team using metals so I'll probably meet some resistance. Can this be enabled globally?

**Reply by tgodzik** [reply]
> Can this be enabled globally?

Unfortunately not, you can create a feature request for that. There might be more people who would prefer it this way, thoughit might be hard to achieve.

> We are using maven (and all the projects have a pom.xml). I see it is possible to add the option there, but I'd need to do that for every repo and I am the only one in my team using metals so I'll probably meet some resistance.

It's a normal linting option that should actually be useful for other people and can be verified on CI. It's good practice to add linting options irrelevant of the project. 

You can see how it's done in maven here: https://davidb.github.io/scala-maven-plugin/example_compile.html under Compiler Arguments

**Reply by dekiesel** [reply]
Thank you. I guess part of my problem is that I am running `scala3 ` locally, which doesn't seem to have the unused-flag

https://docs.scala-lang.org/scala3/guides/migration/options-lookup.html

I couldn't get it working at all.

```
import scala.math.sqrt
val unus = 2
@main def run() =
  val ages = Seq(42, 75, 29, 64)
  println(s"The oldest person is ${ages.max}")
```

```
C:\Users\user>scalac -Wunused:all scalatest.scala

C:\Users\user>
```
I'll consider downgrading or using intellij before pushing. Thank you very much for your help!

**Reply by dekiesel** [reply]
Since you need one flag for 2.12 and another one for 2.13 (and there is no flag for 3) the team declined adding anything to the pom.xml. So for unused imports I am stuck with Intellij.

Thank you for your support again :)
